### DISCUSSION TITLE
Make gossipsub actually listen on 0.0.0.0

### DISCUSSION THREAD

**Discussion by Frederik-Baetens** [libp2p/rust-libp2p#2367] [Q&A] [answered]
`swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?).unwrap();` makes my gossipsub swarm randomly choose to listen on 127.0.0.1 or on my internal ip, but apparently not both. The following code also only returns 127.0.0.1 or my LAN ip:

```
            for addr in swarm.listeners() {
                println!("Listening on {:?}", addr);
                listening = true;
            }
```

 This makes it so that I can't use 127.0.0.1 to connect to libp2p when the swarm decided to only listen on my LAN ip.

With other services, listening on 0.0.0.0, enables listening on both 127.0.0.1 & the LAN/WAN ip. How can I achieve this behavior with libp2p?

**Comment by mxinden** [comment, marked as answer]
I am not sure I follow, nor can I reproduce the issue. Mind elaborating further / adding more details?

Note that `libp2p-gossipsub` as a `NetworkBehaviour` is not aware of the transport protocol, nor which addresses to listen on.

**Reply by Frederik-Baetens** [reply]
I was mistaken about not being able to connect to 127.0.0.1 sometimes, I must have another bug somewhere in my larger project then. However it is annoying that swarm.listeners() randomly switches between returning 127.0.0.1 and my lan ip.

Example here: https://github.com/Frederik-Baetens/libp2p-tokiochat

It's adapted from the examples, and just uses gossipsub without mdns.

You can run it with `cargo run`, and connect a second node with `cargo run -- /ip4/127.0.0.1/tcp/44901` (replace the multiaddr with the one the first node prints.

**Reply by rkuhn** [reply]
If you check right after calling `.listen_on` then you’ll first get one and then the other address, in no particular order. You can observe the Swarm’s events to get notified of all addresses as they are discovered — unfortunately there is no signal from which you can deduce that “now I have all addresses”, because new listen addresses will continue to be discovered as your device’s network environment changes.

**Reply by mxinden** [reply]
To add to what @rkuhn said above, I recommend continuously reporting the listen addresses, e.g. as done here:

https://github.com/libp2p/rust-libp2p/blob/e19391e966d64aad1839aa7781a275c64100c75a/examples/ping.rs#L77-L79
